### PR TITLE
Drain operations should not complete until the new nexus target is connected

### PR DESCRIPTION
**Pull request by chriswldenyer**
Use a vector of drained volumes and the start timestamp of the drain operation so that marking a node as drained can be delayed until no shutdown nexuses remain on that node.

### CODE REVIEW

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 1]
```diff
@@ -7,21 +7,30 @@ use crate::controller::{
...
     transport::{NodeId, RepublishVolume, VolumeId, VolumeShareProtocol},
 };
 
+pub type DrainingVolumes = HashMap<VolumeId, (NodeId, Instant)>;
+
 /// Node drain reconciler.
 #[derive(Debug)]
 pub(super) struct NodeNexusReconciler {
     counter: PollTimer,
+    drainingvolumes: DrainingVolumes,
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: I think we can store these in the node itself, but with serde skip so it's not written to etcd

Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Updated to store the draining volumes in the node spec, unserialized.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 2]
```diff
@@ -72,12 +75,12 @@ async fn republish_volume(
...
 
 /// Drain the specified node if in draining state
-async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> PollResult {
+async fn check_and_drain_node(context: &PollContext, node_spec: &mut NodeSpec) -> PollResult {
     if !node_spec.is_draining() {
         return PollResult::Ok(PollerState::Idle);
     }
 
-    let node_id = node_spec.id();
+    let node_id = node_spec.id().clone();
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Do we need the clone here?

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 3]
```diff
@@ -285,6 +287,19 @@ impl NodeSpec {
         }
     }
 
+    pub fn add_draining_volume(&mut self, volume_id: &VolumeId) {
+        self.draining_volumes
+            .insert(volume_id.clone(), SystemTime::now());
+    }
+
+    pub fn get_draining_volumes(&self) -> HashMap<VolumeId, SystemTime> {
```
Comment (Abhinandan-Purkait) [inline control-plane/stor-port/src/types/v0/store/node.rs]: Should we return a `&HashMap<VolumeId, SystemTime>` to avoid the clones.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 4]
```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
...
+        // If not, the next reconciliation loop will check again.
+        let mut can_set_drained = true;
+
+        let drainingvolumes = context.specs().get_node_draining_volumes(&node_id).await?;
+        for (vi, starttime) in drainingvolumes {
+            let elapsed = starttime.elapsed();
+            let mut remove_vol_from_list = true;
+            if elapsed.is_ok()
+                && elapsed.unwrap() < Duration::from_secs(DRAINING_VOLUME_TIMOUT_SECONDS)
+                && context.specs().volume_rsc(&vi).is_some()
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: You can probably use `if let Some(elapsed) = ` to reduce multiple condition checks.

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 5]
```diff
@@ -183,4 +185,45 @@ impl ResourceSpecsLocked {
         registry.store_obj(&drained_node_spec).await?;
         Ok(drained_node_spec)
     }
+
+    pub(crate) async fn add_node_draining_volume(
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/node/specs.rs]: Can we add some doc comments as the methods are `pub(crate)`. Below as well.

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 6]
```diff
@@ -285,6 +287,19 @@ impl NodeSpec {
         }
     }
 
+    pub fn add_draining_volume(&mut self, volume_id: &VolumeId) {
```
Comment (Abhinandan-Purkait) [inline control-plane/stor-port/src/types/v0/store/node.rs]: Add doc comments for the `pub` methods?

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 4 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: I tried that but rust complains: error[E0658]: `let` expressions in this position are unstable
   --> control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs:172
I'll see if I can find something that works.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 2 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: It seems not, fixed.

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 5 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs]: fixed

File: control-plane/stor-port/src/types/v0/store/node.rs (hunk 6 above)
Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: fixed

File: control-plane/stor-port/src/types/v0/store/node.rs (hunk 3 above)
Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: fixed

Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: Changed to return the reference, but I think a clone() is still needed at the spec layer that calls this function.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 7]
```diff
@@ -32,8 +35,8 @@ impl TaskPoller for NodeNexusReconciler {
         let nodes = context.specs().nodes();
         let mut results = Vec::with_capacity(nodes.len());
 
-        for node in nodes {
-            results.push(check_and_drain_node(context, &node).await);
+        for mut node in nodes {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: why is mut needed here?

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 8]
```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
...
-            .set_node_drained(context.registry(), node_spec.id())
-            .await
-        {
-            tracing::error!(
-                %error,
-                node.id = node_id.as_str(),
-                "Failed to set node to state drained"
-            );
-            return PollResult::Err(error);
+        // All volumes on the node are republished.
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs:221]: can we break this logic into a separate function?

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 9]
```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
...
+                tracing::error!(
+                    %error,
+                    node.id = node_id.as_str(),
+                    "Failed to set node to state drained"
+                );
+                return PollResult::Err(error);
+            }
+            tracing::info!(node.id = node_id.as_str(), "Set node to state drained");
+        } else {
+            tracing::info!(node.id = node_id.as_str(), "Shutdown nexuses remain");
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Can we keep a count?
EDIT: made suggestion above

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 10]
```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
...
+                        nexus.count = shutdown_nexuses.len(),
+                        "Shutdown nexuses remain"
+                    );
+                } // else it is no longer shutdown
+            } // else it is timed out or no longer exists
+            if remove_vol_from_list {
+                tracing::info!(
+                    node.id = node_id.as_str(),
+                    volume.id = vi.as_str(),
+                    "Removing a volume from the draining volume list"
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]:
```suggestion
                    "Removing volume from the draining volume list"
```

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 4 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Can't have multiple if let conditions IIRC, but we might be able to simply do without this check: "context.specs().volume_rsc(&vi).is_some()"
As if the volume is gone, then nexus shutdown list will be empty

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 11]
```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
...
+                );
+                context
+                    .specs()
+                    .remove_node_draining_volume(context.registry(), node_id, &vi)
+                    .await?;
+            }
+        }
+        // Change the node state to "drained"
+        if can_set_drained {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: do we need to keep this variable? 

```suggestion
impl NodeSpec {
fn pending_drain_volumes(&self) -> u64 { self.list_thingy.len() }
}
// ...
match node_spec.pending_drain_volumes() {
  0 => {
  if let Err(error) = context
                .specs()
                .set_node_drained(context.registry(), node_id)
                .await
            {
                tracing::error!(
                    %error,
                    node.id = node_id.as_str(),
                    "Failed to set node to state drained"
                );
                return PollResult::Err(error);
            }
  }
  remaining => { tracing::info!(node.id = node_id.as_str(), "{remaining} Shutdown targets remain"); }
}

```

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 12]
```diff
@@ -183,4 +185,48 @@ impl ResourceSpecsLocked {
...
+        node_id: &NodeId,
+        volume_id: &VolumeId,
+    ) -> Result<NodeSpec, SvcError> {
+        let node = self.node_rsc(node_id)?;
+        let drained_node_spec = {
+            let mut locked_node = node.lock();
+            locked_node.add_draining_volume(volume_id);
+            locked_node.clone()
+        };
+        registry.store_obj(&drained_node_spec).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/specs.rs:202]: hmm do we need to store this in etcd at all?
If so then we should change this, and only write out once to etcd, rather than once per volume add...

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 13]
```diff
@@ -152,8 +152,9 @@ pub struct NodeSpec {
     #[serde(default)] // Ensure backwards compatibility in etcd when upgrading.
     cordon_drain_state: Option<CordonDrainState>,
     #[serde(skip_serializing_if = "Option::is_none")]
-    #[serde(default)] // Ensure backwards compatibility in etcd when upgrading.
     node_nqn: Option<HostNqn>,
+    #[serde(default, skip)] // Ensure backwards compatibility in etcd when upgrading.
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/store/node.rs]: why did you remove serde default from above?
Also seems we're not writing to etcd, so I think we can change previous functions where I commented about etcd.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 14]
```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
...
-                node.id = node_id.as_str(),
-                "Failed to set node to state drained"
-            );
-            return PollResult::Err(error);
+        // All volumes on the node are republished.
+        // Determine whether we can mark the node as drained.
+        // If not, the next reconciliation loop will check again.
+        let mut can_set_drained = true;
+
+        let drainingvolumes = context.specs().get_node_draining_volumes(node_id).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]:
```suggestion
        let draining_volumes = context.specs().get_node_draining_volumes(node_id).await?;
```

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 15]
```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
...
+        for (vi, starttime) in drainingvolumes {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: do we need one timeout per volume, or is one single timeout on the node_spec enough?

File: control-plane/stor-port/src/types/v0/store/node.rs (hunk 13 above)
Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: Looks like that was a typo, fixing....

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 12 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs:202]: True, it doesn't need to be stored.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 8 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs:221]: created new function

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 7 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: It wasn't needed, fixed.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 10 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: fixed

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 12 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs:202]: fixed

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 11 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Changed to use the size of the hashmap.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 4 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: fixed

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 9 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: fixed

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 15 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: yes, true. Could be a single timestamp.

File: control-plane/stor-port/src/types/v0/store/node.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply control-plane/stor-port/src/types/v0/store/node.rs]: Yep, it could be, but its good to return at this point.

**Review by Abhinandan-Purkait — APPROVED** [review]

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 15 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Using single timestamp.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 16]
```diff
@@ -72,20 +75,73 @@ async fn republish_volume(
     Ok(())
 }
 
+async fn find_shutdown_volumes(context: &PollContext, node_id: &NodeId) -> Result<(), SvcError> {
+    let draining_starttime = context.specs().get_node_draining_timestamp(node_id).await?;
+
+    if draining_starttime.is_some() {
+        let elapsed = draining_starttime.unwrap().elapsed();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]:
```suggestion
    if let Some(draining_starttime) = draining_starttime else {
        let elapsed = draining_starttime.elapsed().unwrap_or(DRAINING_VOLUME_TIMOUT_SECONDS);
```
or you can short-circuit
```suggestion
    let Some(draining_starttime) = draining_starttime else {
      return Ok(());
    }
```

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 17]
```diff
@@ -7,11 +7,14 @@ use crate::controller::{
     task_poller::{PollEvent, PollResult, PollTimer, PollerState},
 };
 use agents::errors::SvcError;
+use std::time::Duration;
 use stor_port::types::v0::{
     store::{node::NodeSpec, volume::VolumeSpec},
     transport::{NodeId, RepublishVolume, VolumeId, VolumeShareProtocol},
 };
 
+const DRAINING_VOLUME_TIMOUT_SECONDS: u64 = 120;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]:
```suggestion
const DRAINING_VOLUME_TIMEOUT_SECONDS: u64 = 120;
```

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 18]
```diff
@@ -136,6 +192,19 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
                         node.id = node_spec.id().as_str(),
                         "Moved volume"
                     );
+                    if let Err(error) = context
+                        .specs()
+                        .add_node_draining_volume(context.registry(), node_spec.id(), &vol_id)
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Can we split this loop into 2, and set all draining volumes in 1 go to avoid taxing etcd?

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 19]
```diff
@@ -183,4 +185,94 @@ impl ResourceSpecsLocked {
...
+    }
+
+    /// Get the vector of draining volumes on this node.
+    pub(crate) async fn get_node_draining_volumes(
+        &self,
+        node_id: &NodeId,
+    ) -> Result<Vec<VolumeId>, SvcError> {
+        let node = self.node_rsc(node_id)?;
+        let locked_node = node.lock().clone();
+        Ok(locked_node.get_draining_volumes())
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/specs.rs]: don't need to clone the node, but rather the draining_volumes only

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 20]
```diff
@@ -183,4 +185,94 @@ impl ResourceSpecsLocked {
...
+            let mut locked_node = node.lock();
+            locked_node.add_draining_volume(volume_id);
+            locked_node.clone()
+        };
+        registry.store_obj(&drained_node_spec).await?;
+        Ok(drained_node_spec)
+    }
+
+    /// Get the vector of draining volumes on this node.
+    pub(crate) async fn get_node_draining_volumes(
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/specs.rs]:
```suggestion
    pub(crate) async fn node_draining_volumes(
```
getters in rust don't tend to have "get_" prefix

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 21]
```diff
@@ -183,4 +185,94 @@ impl ResourceSpecsLocked {
...
+    }
+
+    /// Set the draining timestamp on this node.
+    pub(crate) async fn set_draining_timestamp_if_none(
+        &self,
+        node_id: &NodeId,
+    ) -> Result<(), SvcError> {
+        let node = self.node_rsc(node_id)?;
+        let mut locked_node = node.lock().clone();
+        locked_node.set_draining_timestamp_if_none();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/specs.rs]: hmm, this is setting the timestamp on a clone and not the node itself?

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 22]
```diff
@@ -183,4 +185,94 @@ impl ResourceSpecsLocked {
...
+        Ok(())
+    }
+
+    /// Get the draining timestamp on this node.
+    pub(crate) async fn get_node_draining_timestamp(
+        &self,
+        node_id: &NodeId,
+    ) -> Result<Option<SystemTime>, SvcError> {
+        let node = self.node_rsc(node_id)?;
+        let locked_node = node.lock().clone();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/specs.rs]: is this clone needed?

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 23]
```diff
@@ -154,6 +154,10 @@ pub struct NodeSpec {
     #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(default)] // Ensure backwards compatibility in etcd when upgrading.
     node_nqn: Option<HostNqn>,
+    #[serde(default)] // Ensure backwards compatibility.
+    draining_volumes: Vec<VolumeId>,
+    #[serde(default, skip)] // Ensure backwards compatibility and do not store.
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/store/node.rs]: Don't think you need default here as we have skip?

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 24]
```diff
@@ -343,6 +349,50 @@ impl NodeSpec {
...
+    /// Add the draining volume to the node spec for checking shutdown nexuses.
+    pub fn add_draining_volume(&mut self, volume_id: &VolumeId) {
+        self.draining_volumes.push(volume_id.clone());
+        if self.draining_timestamp.is_none() {
+            self.draining_timestamp = Some(SystemTime::now())
+        }
+    }
+
+    /// Get the vector of draining volumes on this node.
+    pub fn get_draining_volumes(&self) -> Vec<VolumeId> {
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/store/node.rs]:
```suggestion
    pub fn draining_volumes(&self) -> Vec<VolumeId> {
```
and below on other get_**

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 17 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: fixed

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 20 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs]: fixed

File: control-plane/stor-port/src/types/v0/store/node.rs (hunk 24 above)
Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: fixed

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 19 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs]: fixed

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 22 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs]: fixed

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 21 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs]: fixed

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 16 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: fixed using the short-circuit

File: control-plane/stor-port/src/types/v0/store/node.rs (hunk 23 above)
Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: fixed

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 18 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: Took add_node_draining_volume() out of the loop. Made a similar change in find_shutdown_volumes().
It now adds/removes sets of volumes by using HashSet as the parameter.
The volume ids are now stored as a HashSet which removes the risk of duplicating volume ids in the vector.

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 25]
```diff
@@ -72,20 +75,77 @@ async fn republish_volume(
...
+            let shutdown_nexuses = context
+                .registry()
+                .specs()
+                .volume_shutdown_nexuses(&vi)
+                .await;
+            if !shutdown_nexuses.is_empty() {
+                // if it still has shutdown nexuses
+                tracing::info!(
+                    node.id = node_id.as_str(),
+                    volume.id = vi.as_str(),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]:
```suggestion
                    volume.uuid = vi.as_str(),
```

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs [hunk 26]
```diff
@@ -72,20 +75,77 @@ async fn republish_volume(
...
+                    nexus.count = shutdown_nexuses.len(),
+                    "Shutdown nexuses remain"
+                );
+            } else {
+                tracing::info!(
+                    node.id = node_id.as_str(),
+                    volume.id = vi.as_str(),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]:
```suggestion
                    volume.uuid = vi.as_str(),
```

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 27]
```diff
@@ -183,4 +185,94 @@ impl ResourceSpecsLocked {
...
+        Ok(locked_node.draining_volumes())
+    }
+
+    /// Get the number of draining volumes on this node.
+    pub(crate) async fn node_draining_volume_count(
+        &self,
+        node_id: &NodeId,
+    ) -> Result<usize, SvcError> {
+        let node = self.node_rsc(node_id)?;
+        let locked_node = node.lock().clone();
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/specs.rs]: don't need to clone?

File: control-plane/agents/src/bin/core/node/specs.rs [hunk 28]
```diff
@@ -183,4 +185,94 @@ impl ResourceSpecsLocked {
         registry.store_obj(&drained_node_spec).await?;
         Ok(drained_node_spec)
     }
+
+    /// Add the draining volume to the node spec for checking shutdown nexuses.
+    pub(crate) async fn add_node_draining_volumes(
+        &self,
+        registry: &Registry,
+        node_id: &NodeId,
+        draining_volumes: &HashSet<VolumeId>,
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/specs.rs]: I think we can consume by value here?

**Review by tiagolobocastro — APPROVED** [review]

**Comment by tiagolobocastro** [comment]
@chriswldenyer please squash the commits before merging

**Review by Abhinandan-Purkait — APPROVED** [review]

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 25 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: fixed

File: control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs (hunk 26 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs]: fixed

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 27 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs]: fixed

File: control-plane/agents/src/bin/core/node/specs.rs (hunk 28 above)
Comment (chriswldenyer) [reply control-plane/agents/src/bin/core/node/specs.rs]: Fixed. Did the same for remove_node_draining_volumes().

**Comment by chriswldenyer** [comment]
bors try

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 29]
```diff
@@ -227,20 +236,24 @@ impl NodeSpec {
...
             Some(ds) => {
                 ds.add_cordon_label(&label);
-                self.resolve();
             }
             None => {
                 //add the label and set the state to cordoned
                 self.cordon_drain_state = Some(CordonDrainState::cordon(&label));
             }
         }
+        self.resolve();
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/store/node.rs:255]: what's this resolve thing?

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 30]
```diff
@@ -343,6 +357,52 @@ impl NodeSpec {
             None => false,
         }
     }
+
+    /// Add the draining volumes to the node spec for checking shutdown nexuses.
+    pub fn add_draining_volumes(&mut self, draining_volumes: HashSet<VolumeId>) {
+        if !draining_volumes.is_empty() {
+            self.draining_volumes
+                .extend(draining_volumes.iter().cloned());
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/store/node.rs]:
```suggestion
                .extend(draining_volumes.into_iter());
```

Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: fixed

**Review by Abhinandan-Purkait — APPROVED** [review]

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/stor-port/src/types/v0/store/node.rs [hunk 31]
```diff
@@ -227,20 +236,24 @@ impl NodeSpec {
             },
             None => {}
         }
+        match &mut self.cordon_drain_state {
+            Some(CordonDrainState::Draining(_)) => {}
+            _ => self.remove_all_draining_volumes(),
+        }
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/store/node.rs]:
```suggestion
        if !matches!(self.cordon_drain_state, Some(CordonDrainState::Draining(_))) {
            self.remove_all_draining_volumes();
        }
```

**Review by tiagolobocastro — APPROVED** [review]
Is this ready to merge?

File: control-plane/stor-port/src/types/v0/store/node.rs (hunk 31 above)
Comment (chriswldenyer) [reply control-plane/stor-port/src/types/v0/store/node.rs]: fixed

**Comment by chriswldenyer** [comment]
bors try

**Comment by chriswldenyer** [comment]
bors merge

**Comment by chriswldenyer** [comment]
bors merge

**Comment by chriswldenyer** [comment]
bors merge-

**Comment by chriswldenyer** [comment]
bors merge
